cmake/BDWGC.cmake

```diff
@@ -0,0 +1,74 @@
+macro(p4c_obtain_bdwgc)
+  include(CheckSymbolExists)
+
+  option(
+    P4C_USE_PREINSTALLED_BDWGC
+    "Look for a preinstalled version of BDWGC in the system instead of installing a prebuilt binary using FetchContent."
+    OFF
+  )
+  set(LIBGC_LIBRARIES gc gccpp cord)
+  set(HAVE_LIBGC 1)
+
+  if(P4C_USE_PREINSTALLED_BDWGC)
+    OPTION (ENABLE_MULTITHREAD "Use multithreading" OFF)
+    find_package(LibGc 7.2.0 REQUIRED)
+    check_function_exists (GC_print_stats HAVE_GC_PRINT_STATS)
+  else()
+    # Print out download state while setting up BDWGC.
+    set(FETCHCONTENT_QUIET_PREV ${FETCHCONTENT_QUIET})
+    set(BUILD_SHARED_LIBS_PREV ${BUILD_SHARED_LIBS})
+    set(FETCHCONTENT_QUIET OFF)
+    set(BUILD_SHARED_LIBS OFF CACHE BOOL "")
+
+    # Turn on C++ support.
+    set(enable_cplusplus ON CACHE BOOL "C++ support")
+    set(install_headers OFF CACHE BOOL "Install header and pkg-config metadata files.")
+    set(enable_docs ${ENABLE_DOCS} CACHE BOOL "Build and install documentation.")
+    # We may have to compile very large programs which allocate a lot on the heap.
+    set(enable_large_config ON CACHE BOOL "Optimize for large heap or root set.")
+    set(without_libatomic_ops OFF CACHE BOOL "Use atomic_ops.h in libatomic_ops/src")
+    # Redirect all malloc calls in the program.
+    set(enable_redirect_malloc OFF CACHE BOOL "Redirect malloc and friends to GC routines.")
+    # Try to enable thread-local-storage for better performance.
+    set(enable_thread_local_alloc ON CACHE BOOL "Turn on thread-local allocation optimization")
+
+    fetchcontent_declare(
+      bdwgc
+      GIT_REPOSITORY https://github.com/ivmai/bdwgc.git
+      GIT_TAG e340b2e869e02718de9c9d7fa440ef4b35785388 # 8.2.6
+      GIT_PROGRESS TRUE
+      PATCH_COMMAND
+        git apply ${P4C_SOURCE_DIR}/cmake/bdwgc.patch || git apply
+        ${P4C_SOURCE_DIR}/cmake/bdwgc.patch -R --check && echo
+        "Patch does not apply because the patch was already applied."
+    )
+    fetchcontent_makeavailable_but_exclude_install(bdwgc)
+
+    # Bdwgc source code may trigger warnings which we need to ignore.
+    set_target_properties(gc PROPERTIES COMPILE_FLAGS "-Wno-error -w")
+    set_target_properties(gccpp PROPERTIES COMPILE_FLAGS "-Wno-error -w")
+    set_target_properties(cord PROPERTIES COMPILE_FLAGS "-Wno-error -w")
+
+    # Add some extra compile definitions which allow us to use our customizations.
+    target_compile_definitions(gc PUBLIC
+      HANDLE_FORK ENABLE_DISCLAIM GC_USE_DLOPEN_WRAP SKIP_CPP_DEFINITIONS
```

Comment:
Will do. 
For context on `DLOPEN_WRAP` https://github.com/ivmai/bdwgc/blob/master/docs/README.macros#L416 and https://github.com/ivmai/bdwgc/issues/561#issuecomment-1610830189
It seems to help in multi-threaded environments. 

For an explanation of `SKIP_CPP_DEFINTIONS` see https://github.com/p4lang/p4c/pull/3930#discussion_r1554197842

Reply:
> > GC_USE_DLOPEN_WRAP      Causes the collector to redefine malloc and
> > intercepted pthread routines with their real names, and causes it to use
> > dlopen and dlsym to refer to the original versions.  This makes it possible
> > to build an LD_PRELOADable malloc replacement library.
> 
> Doesn't this interfere with our own re-declaration of `malloc`? Or maybe the BDWGC one is not linked because there is already malloc defined by our binary...

I think it would be ok, as malloc ends up redirected to the gc regardless, but the main reason we have redeclarations of malloc is that the default libgc builds on Ubuntu and others DON'T use GC_USE_DLOPEN_WRAP.  With it, we should be able to remove our malloc replacements.

Reply:
yes, I was wondering if we could actually get rid of our `gc.cpp` (we would have to be willing to drop support for distribution version of libgc in most distros).

Reply:
@vlstill I'm not sure we can drop on Darwin. Though, it's worth investigating.